Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+about Kubernetes' scheduling of Ceph daemons.
+
+Kubernetes has two Resource Request types: Requests and Limits. Requests govern scheduling, and
+Limits instruct Kubernetes to kill and restart application Pods when they are over-consuming given
+Limits.
+
+When there are Ceph hardware requirements, treat those requirements as Requests, not Limits. This is
+because all Ceph daemons are critical for storage, and it is best to never set Resource Limits for
+Ceph Pods. If Ceph Daemons are over-consuming Requests, there is likely a failure scenario
+happening. In a failure scenario, killing a daemon beyond a Limit is likely to make an already bad
+situation worse. This could create a “thundering herds” situation where failures synchronize and
+magnify.
+
+Generally, storage is given minimum resource guarantees, and other applications should be limited so
+as not to interfere. This guideline already applies to bare-metal storage deployments, not only for
+Kubernetes.
+
+As you read on, it is important to note that all recommendations can be affected by how Ceph daemons
+are configured. For example, any configuration regarding caching. Keep in mind that individual
+configurations are out of scope for this document.
+
+### Resource Requests - MON/MGR
+
+Resource Requests for MONs and MGRs are straightforward. MONs try to keep memory usage to around
+1GB; however, that can expand under failure scenarios. We recommend 4 GB RAM and 4 CPU cores.
+
+Recommendations for MGR nodes are harder to make, since enabling more modules means higher usage. We
+recommend starting with 2 GB RAM and 2 CPU cores for MGRs. It is a good idea to look at the actual
+usage for Deployments, and do not forget to consider usage during failure scenarios.
+
+MONs:
+ * Request 4 CPU cores
+ * Request 4GB RAM (2.5GB minimum)
+
+MGR:
+ * Memory will grow the more MGR modules are enabled
+ * Request 2 GB RAM and 2 CPU cores
+
+### Resource Requests - OSD CPU
+
+Recommendations and calculations for OSD CPU are straightforward.
+
+Hardware recommendations:
+ * 1 x 2GHz CPU Thread per spinner
+ * 2 x GHz CPU Thread per SSD
+ * 4 x GHz CPU Thread per NVMe
+
+Examples:
+ * 8 HDDS journaled to SSD – `10 cores / 8 OSDs = 1.25 cores per OSD`
+ * 6 SSDs without journals – `12 cores / 6 OSDs = 2 cores per OSD`
+ * 8 SSDs journaled to NVMe – `20 cores / 8 OSDs = 2.5 cores per OSD`
+
+Note that resources are applied cluster-wide to all OSDs. If a cluster contains multiple OSD types,
+you must use the highest Requests for the whole cluster. For the examples below, a mixture of HDDs
+journaled to SSD and SSDs without journals would necessitate a Request for 2 cores.
+
+### Resource Requests - OSD RAM
+
+We have node hardware recommendations for OSD RAM usage, and we need to translate this to RAM
+requests on a per-OSD basis. The node-level recommendation we have is given below where
+`osd_memory_target` is a Ceph configuration described more afterwards.
```

Comment:
I'd consider a rephrase here, I see why you've added what you have - but it doesn't flow well.

_"The node-level recommendation below describes `osd_memory_target`; a Ceph configuration that is described in detail afterwards."_

Now you've got me on the semi-colon train

Reply:
But I don't see why the original wasn't better? _"The node-level recommendation below describes osd_memory_target. This is a Ceph configuration that will be described further on."_